compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs

```diff
@@ -743,8 +743,166 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
+                                can_suggest_clone = false;
+                            }
+                        }
+                        _ => {}
+                    }
+                }
+                hir::ExprKind::Loop(..) => {
+                    outer_most_loop = Some(e);
+                }
+                _ => {}
+            }
+        }
+
+        /// Look for `break` expressions within any arbitrary expressions. We'll do this to infer
+        /// whether this is a case where the moved value would affect the exit of a loop, making it
+        /// unsuitable for a `.clone()` suggestion.
+        struct BreakFinder {
+            found_break: bool,
+        }
+        impl<'hir> Visitor<'hir> for BreakFinder {
+            fn visit_expr(&mut self, ex: &'hir hir::Expr<'hir>) {
+                if let hir::ExprKind::Break(..) = ex.kind {
+                    self.found_break = true;
+                }
+                hir::intravisit::walk_expr(self, ex);
+            }
+        }
+        if let Some(in_loop) = outer_most_loop
+            && let Some(parent) = parent
+            && let hir::ExprKind::MethodCall(..) | hir::ExprKind::Call(..) = parent.kind
+        {
+            // FIXME: We could check that the call's *parent* takes `&mut val` to make the
+            // suggestion more targeted to the `mk_iter(val).next()` case. Maybe do that only to
+            // check for wheter to suggest `let value` or `let mut value`.
+
+            let span = in_loop.span;
+            let mut finder = BreakFinder { found_break: false };
+            finder.visit_expr(in_loop);
+            let sm = tcx.sess.source_map();
+            if (finder.found_break || true)
+                && let Ok(value) = sm.span_to_snippet(parent.span)
+            {
+                // We know with high certainty that this move would affect the early return of a
+                // loop, so we suggest moving the expression with the move out of the loop.
+                let indent = if let Some(indent) = sm.indentation_before(span) {
+                    format!("\n{indent}")
+                } else {
+                    " ".to_string()
+                };
+                err.multipart_suggestion(
+                    "consider moving the expression out of the loop so it is only moved once",
+                    vec![
+                        (parent.span, "value".to_string()),
+                        (span.shrink_to_lo(), format!("let mut value = {value};{indent}")),
+                    ],
+                    Applicability::MaybeIncorrect,
+                );
+            }
+        }
+        can_suggest_clone
+    }
+
     fn suggest_cloning(&self, err: &mut Diag<'_>, ty: Ty<'tcx>, expr: &hir::Expr<'_>, span: Span) {
         let tcx = self.infcx.tcx;
+        if !self.suggest_hoisting_call_outside_loop(err, expr) {
+            // The place where the the type moves would be misleading to suggest clone. (#121466)
+            return;
+        }
```

Comment:
This is doing something different than what "suggest_cloning" suggests. Could you move the call to `suggest_hoisting_call_outside_loop` near the call site of `suggest_cloning`, or rename `suggest_cloning` to something like `suggest_cloning_or_hoisting`